File: drun/drun/k8s.py
#
#    Copyright 2017 EPAM Systems
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.
#
"""
DRun k8s functions
"""
import logging
import os

import drun
import drun.env
import drun.headers

import kubernetes
import kubernetes.client
import kubernetes.config
import kubernetes.config.config_exception

K8S_LABEL_PREFIX = 'drun/'
K8S_LOCAL_CLUSTER_DOMAIN = 'cluster.local'


def build_client():
    """
    Configure and returns kubernetes client

    :return: :py:module:`kubernetes.client`
    """
    try:
        kubernetes.config.load_incluster_config()
    except kubernetes.config.config_exception.ConfigException:
        kubernetes.config.load_kube_config()

    return kubernetes.client.ApiClient()


def get_service_url(service, port_name):
    """
    Get service's url for specific port name

    :param service: service object
    :type service: :py:class:`kubernetes.client.models.v1_service.V1Service`
    :param port_name: name of port
    :type port_name: str
    :return: str -- url to service's port
    """
    ports = {port.name: port for port in service.spec.ports}
    if port_name not in ports:
        raise Exception('Cannot found port named %s in %s service' % (port_name, service.metadata.name))

    url = '%s.%s.svc.%s:%d' % (service.metadata.name,
                               service.metadata.namespace,
                               K8S_LOCAL_CLUSTER_DOMAIN,
                               ports[port_name].port)
    return url


def find_service(service_name, deployment, namespace='default'):
    """
    Find service in cluster

    :param service_name: service name
    :type service_name: str
    :param deployment: deployment name
    :type deployment: str
    :param namespace: namespace
    :type namespace: str
    :return: :py:class:`kubernetes.client.models.v1_service.V1Service` or None
    """
    client = build_client()

    core_api = kubernetes.client.CoreV1Api(client)
    all_services = core_api.list_namespaced_service(namespace)

    valid_services = [
        service
        for service in all_services.items
        if service.metadata.name == '%s-%s' % (deployment, service_name)
    ]

    if len(valid_services) > 1:
        raise Exception('Founded more that one valid service for %s' % service_name)

    if len(valid_services) == 0:
        raise Exception('Cannot found valid service for %s' % service_name)

    return valid_services[0]

File: legion/operator/pkg/utils/k8s.go
//
//    Copyright 2019 EPAM Systems
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.
//

package utils

import (
	"io"
	"net/http"

	core_v1 "k8s.io/api/core/v1"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/rest"
	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
)

var logk8s = logf.Log.WithName("legion_k8s")

type Writer interface {
	http.Flusher
	http.CloseNotifier
	io.Writer
}

// Stream logs from the pod to the writer object
// k8sConfig     - kubernetes config
// namespace     - namespace where pod is located
// podName       - kubernetes pod name
// containerName - container name of the pod
// writer        - logs will be streamed into this writer
// follow        - specify if the logs should be streamed
// logFlushSize  - size of the flushed logs per chunk
func StreamLogs(
	k8sConfig *rest.Config, namespace string, podName string,
	containerName string, writer Writer, follow bool, logFlushSize int64,
) error {
	clientset, _ := kubernetes.NewForConfig(k8sConfig)
	request := clientset.CoreV1().Pods(namespace).
		GetLogs(podName, &core_v1.PodLogOptions{Follow: follow, Container: containerName})

	reader, err := request.Stream()
	if err != nil {
		logk8s.Error(err, "open log stream")
		return err
	}
	defer reader.Close()

	clientGone := writer.CloseNotify()
	for {
		select {
		case <-clientGone:
			return nil
		default:
			_, err := io.CopyN(writer, reader, logFlushSize)
			if err != nil {
				if err == io.EOF {
					return nil
				}

				return err
			}

			writer.Flush()
		}
	}
}
